### ISSUE TITLE
Pull request folder structure view about all files changed

### ISSUE THREAD

**Issue by xserrat** [dear-github/dear-github#195]
It would be nice to put on the PR section, inside `files changed`, an alternative view in order to show files as a folder structure and navigate through it to find each file more easily.

Thanks and good work! 👍

**Comment by Mottie** [comment]
You can click on the "# files" dropdown to view the full list

![2016-09-16 21_20_17-core_ deprecate jquery isarray by owsmmj pull request 3278 jquery_jquery](https://cloud.githubusercontent.com/assets/136959/18605294/80213fda-7c53-11e6-83f9-e3adf29892b1.png)

I also have two userscripts that make navigating the files a little easier:

* https://greasyfork.org/en/scripts/18788-github-diff-file-toggle
* https://greasyfork.org/en/scripts/21559-github-diff-links

**Comment by xserrat** [comment]
Thanks @Mottie to comment my proposal. I'd seen the "# files" dropdown functionality, but in order to preview the files changed, my idea is about a folder structure like this:

<img src="https://cloud.githubusercontent.com/assets/5933658/18618702/4ab330c0-7dec-11e6-8273-c3baf634a88e.jpg" width=200px">

So, when you go to the files changed, you'll see using folder structure view the files changed following the project structure and as a code reviewer you could tell to the pull requester things like: _this file would be better into another folder..._. 

However, thanks for the answer with these userscripts!

**Comment by intellix** [comment]
Bitbucket have had this since the beginning. When looking at a PR you see a folder structure on the left like you do in an IDE. It's one of the pieces majorly missing here I think.
You can use that dropdown, but then you're throw down the page upon clicking one of em

**Comment by xserrat** [comment]
Thanks for your comment @intellix , I have the same idea. I'll be looking forward to this feature 👍

**Comment by Jckf** [comment]
I support this request. Some of our PRs can get quite big, and a tree view of the changed files would give a much nicer overview.

**Comment by intellix** [comment]
Also to add, right now I'm doing a PR on a rather large PR (88 files) and the UI becomes almost unusable due to poor performance. If only individual files were displayed I'm sure it would be much faster

**Comment by kevdoran** [comment]
A file tree picker for diffs, [similar to Bitbucket](https://confluence.atlassian.com/bitbucketserver/review-and-discuss-a-pull-request-808488540.html), would be a welcome feature! 

In the meantime, perhaps a Chrome extension could provide this functionality? Does anyone know of one that already exists?

**Comment by frankkoornstra** [comment]
A tree view would be awesome!

**Comment by picheli20** [comment]
@kevdoran I've just create a chrome extension https://chrome.google.com/webstore/detail/github-code-tree/caffoilocbbhnkdlpdcfanpbeholjdii

I still improving it, so if someone have some suggestion: https://github.com/picheli20/github-code-tree

**Comment by xserrat** [comment]
Wow @picheli20 ! That's a great chrome extension! It works really good!! :) Congrats!!

**Comment by Mottie** [comment]
Actually, [Octotree](https://chrome.google.com/webstore/detail/octotree/bkhaagjahfmjljalopjnoealnfndnagc) added something similar just recently ([ref](https://github.com/buunguyen/octotree/pull/435)).

![](https://user-images.githubusercontent.com/3138607/29946048-97effcc4-8e58-11e7-97ce-0cebb24ed496.png)

**Comment by kevdoran** [comment]
Yep, I've been using Octotree as well, @Mottie, which works well and would recommend it. It really improves the GitHub repo browsing experience, both inside and outside the context of a PR.

@picheli20 - I just installed and tried out your github-code-tree extension. Excellent start! One thing I noticed is that clicking on a file in github-code-tree jumps to the correct position in the main diff view much faster than Octotree, at least at the time of this writing. Granted, Octotree only recently added the PR-view capabilties (it started as a file tree viewer for general repo browsing), so maybe that performance will improve over time.

One improvement I would make to GitHub Code Tree is auto-expanding folders if they only contain one child folder. For Java projects that create nested folders to math the package structure, 6+ folders can by typical until you get to a code file, and expanding each one can be tedious.

Both are great extensions that improve the PR browsing experience.

**Comment by xserrat** [comment]
Hi @Mottie and @kevdoran ,

That's another nice and useful chrome extension to improve our experience on Github. 

Thank you so much for your contributions, I hope Github hears us including these kinds of features!

**Comment by HerbertBodner** [comment]
Awesone chrome extension @picheli20 ! Congrats
and +1 for adding the tree view feature to the github website

**Comment by rjhilgefort** [comment]
Bump. Would love to see this feature! Has anyone heard if there are any plans to do this?

**Comment by borNfreee** [comment]
One of the most used feature in Bitbucket during a review. Not clear why Github still doesn't have it.

**Comment by CannibalVox** [comment]
It's particularly frustrating when a golang PR has a lot of vendoring changes.  It's not always great to have to split these into two PRs to make them legible, since they may not build individually once separated.

**Comment by jeremejevs** [comment]
Octotree doesn't work for large PRs, unfortunately: https://github.com/buunguyen/octotree/issues/504

**Comment by chris-dickson** [comment]
+1 for the plugin from @picheli20.  Octotree requires an access token for private repositories which I don't feel comfortable with. GitHub Code Tree does everything on the fly with jQuery and regular expressions.

**Comment by picheli20** [comment]
@kevdoran I've just added the option to merge empty folders and some others config :)
<img width="226" alt="screen shot 2018-07-23 at 14 19 41" src="https://user-images.githubusercontent.com/3877606/43076127-7aa3c8a4-8e83-11e8-887d-d18c7caef4d6.png">

**Comment by kevdoran** [comment]
@picheli20 nice work! Thanks for implementing that feature

**Comment by PropertyOfTuckerSystems** [comment]
Sorry if this is going to sound tough but I have to voice the opinion many keep to themselves:
GitHub's PR UX is an absolute usability disaster, I have yet to see anything worse (ok maybe review board) and it gets like 0 attention. Sorry to break the news, but we don't need animoji reactions to all the irrelevant items on the site, we need _actually working pull requests_ - comment threads, collapsible comments, comment actions (won't fix, resolved, pending, etc.), file tree view with an option to see other unmodified files, basic code navigation. All of these are contributing factors to better code quality and spotting issues in PRs, this brings *an actual value* which is hard to overestimate. VSTS, Bitbucket/Stash and some others have been doing this for ages, if I were to suggest a hosting platform, GitHub would not be in my top 5 only because of the PR experience, collaboration is all about PRs and efficient reviews which is totally broken here.

**Comment by GuyKroizman** [comment]
You can collapse all diffs in a pull request which makes it easier to see the folders.
https://blog.github.com/changelog/2018-08-17-collapse-all-diffs-in-a-pull-request-at-once/

**Comment by PeledYuval** [comment]
+1 for this suggestion.
I just moved from a project using Bitbucket to a project using Github and I'm very sad to say that Bitbucket has a much better pull request experience.
Sadly I don't trust user-made extensions when working on company proprietary code.

**Comment by sarachai** [comment]
Another extremely huge +1 for this... Former Atlassian Stash (formerly Bitbucket) user here and I feel like I have no idea what's going on in large PRs with the current UI.

**Comment by mrworkman** [comment]
This is a sorely needed feature. I don't like bitbucket much, but its file tree helps make large PRs easier to wrap your (my) head around.

**Comment by arthakdoo** [comment]
Microsoft has bought GitHub so let's better stop hoping they will improve it...

**Comment by arthakdoo** [comment]
Thanks for creating GitHub Code Tree, unfortunately for me it doesn't display all files. I have also tried GitHub PR Tree (https://github.com/berzniz/github_pr_tree), which works much better (at least for me). Unified view and then maximize icon, works like a charm!

**Comment by juanmendes** [comment]
gitlab has the same feature and I find it extremely useful. That's what we use internally at VMware. 
![image](https://user-images.githubusercontent.com/549331/78203872-1b327800-7466-11ea-8b30-a43820a585f7.png)

**Comment by bartread** [comment]
I agree completely. We're moving towards a mono-repo structure which has massive advantages for us when it comes to development and debugging workflow, but makes it increasingly challenging to see what has been changed within a release of a particular component or service, which we still prefer to be able to deploy separately.

A folder view that would allow you to browse and inspect changes made within individual folders and files would be of huge benefit here.

For example, right now I'm trying to understand what has been changed in one of our sites that has just been deployed to our test environment, but am really struggling to make headway because there's no way for me to browse changes across the multiple pull requests within this release in a way that makes sense conceptually (files and folders).

**Comment by w84miracle** [comment]
+1 for this feature,
Able to skip dependencies/lib directories while reviewing PR would be a great help.

**Comment by jennypns** [comment]
+1 My team would love this feature.

I'm surprised Github doesn't have this feature, especially since Bitbucket and Gitlab both have it.

The github_pr_tree plugin (https://github.com/berzniz/github_pr_tree) works great so far.

**Comment by roman-vi** [comment]
+1 to that issue from our team. Unfortunately it's not in GitHub roadmap https://github.com/github/roadmap/issues?q=is%3Aissue+is%3Aopen

**Comment by asandler** [comment]
Dear Lord, may the voices of people be heard! We are craving for this feature!

**Comment by ndrsllwngr** [comment]
I believe a native folder structure view (code tree) in the scope of pull requests would significantly improve the developer experience. From my experience, the information provided by this feature helps to grasp the impact of a PR. Further, it gives context while reviewing changes which are otherwise outside of the viewport and allows navigating the code more easily.

GitHub already has plenty of excellent features to support PR reviews. However, the lack of this fundamental visual representation of a PR feels rather significant. Hopefully, GitHub will add it to its roadmap soon, since I have grown accustomed to it using Bitbucket and GitLab.

**Comment by julianeden** [comment]
Everything in @troshko111 's comment is SPOT ON. This thread four years old and none of these changes have been implemented. :(

**Comment by chiggans** [comment]
+1 for this feature.

**Comment by Deishelon** [comment]
I just can't believe this is not in the roadmap 😨

**Comment by ohierro-verimatrix** [comment]
Seriously, Bitbucket is like 1000 times better on this front. Get your act together, Github!

**Comment by willvictor** [comment]
+1 for this feature.

**Comment by alexpenedo** [comment]
+1 for this feature

**Comment by massimilianog** [comment]
+1 for this feature

**Comment by akar-shan** [comment]
+1

**Comment by mjakobsen** [comment]
+1 for this as well

**Comment by jsubirat** [comment]
A big +1 here, I really miss this feature so much, as I was used to have it when I used Bitbucket in my previous company...

**Comment by mk185283** [comment]
> Sorry if this is going to sound tough but I have to voice the opinion many keep to themselves:
> GitHub's PR UX is an absolute usability disaster, I have yet to see anything worse (ok maybe review board) and it gets like 0 attention. Sorry to break the news, but we don't need animoji reactions to all the irrelevant items on the site, we need _actually working pull requests_ - comment threads, collapsible comments, comment actions (won't fix, resolved, pending, etc.), file tree view with an option to see other unmodified files, basic code navigation. All of these are contributing factors to better code quality and spotting issues in PRs, this brings _an actual value_ which is hard to overestimate. VSTS, Bitbucket/Stash and some others have been doing this for ages, if I were to suggest a hosting platform, GitHub would not be in my top 5 only because of the PR experience, collaboration is all about PRs and efficient reviews which is totally broken here.

Seconded. It's a shame such a great tool hasn't solved this yet.

**Comment by me250225** [comment]
Our company migrated recently all repos to GitHub from Bitbucket. We lived in assumption that UX will be better in overall. Four years without progress on this request, really?! GitHub c'mon...

**Comment by muodov** [comment]
@me250225 this is not an official github issue tracker, so they don't really have to address this I guess, and it's not clear whether this is _actually_ among the most needed feature. The fact that we need it, doesn't mean the majority of other users do.
In the meantime, there are a few great browser extensions that work pretty well. For example, https://github.com/berzniz/github_pr_tree and https://www.octotree.io/.

**Comment by me250225** [comment]
> @me250225 this is not an official github issue tracker, so they don't really have to address this I guess, and it's not clear whether this is _actually_ among the most needed feature. The fact that we need it, doesn't mean the majority of other users do.
> In the meantime, there are a few great browser extensions that work pretty well. For example, https://github.com/berzniz/github_pr_tree and https://www.octotree.io/.

IMHO, GitHub should have this implemented natively already - take a peek into Bitbucket or GitLab for instance - they do offer such a feature..

Nevertheless, thanks for the tips. We may as well try them out.

**Comment by JamesAnthonyPino** [comment]
Just wanted to leave my vote for this as well. This is a much needed usability improvement. BitBucket has a much more intuitive PR experience and I have never felt lost reviewing even very large PRs. With GitHub, I immediately feel lost since I have no context of where the file is supposed to be in relation to the rest of the code. Context is extremely important! A tree/folder view is _also_ how most developers are accustomed to seeing code in IDEs.

**Comment by GerardRodes** [comment]
This is needed.
I expect Github to implement it on the future, for the moment I'll install the chrome extension

**Comment by Denijar** [comment]
2021 and GitHub still doesn't have this feature. In React projects (as one example of many), viewing and reviewing if the file structure matches project convention's is extremely important. It is almost impossible to do this with GitHub.

**Comment by starikcetin** [comment]
This would be useful. Currently I am doing a local merge with a Git GUI in order to see the changes in a folder view. Lots of time wasted.

**Comment by andrewlehmann-whs** [comment]
Please, Github. After moving from Bitbucket to this site as a result of switching jobs, the PR review experience here is miserable. Even more so because I know how much better it could be.

**Comment by ibaguio** [comment]
+1 Coming from bitbucket, this feature is a must have when reviewing PRs. Imagine having to go up and down the files view and trying to find a file when you have a handful of files changed.

**Comment by ikt32** [comment]
Yes please - it's a chore reviewing larger pull requests, especially those where large swaths of "third party" code is added/removed.

**Comment by nodehack** [comment]
Bump!!!

No File Structure View, TOC doesn't work on large pull requests, and no comment on file features.  Coming from BitBucket projects I feel like I'm back in the stone ages.  Get this on your roadmap!

**Comment by hanhannan** [comment]
must needed feature. its a cool feature, bitbucket has it.

**Comment by BartlomiejSkwira** [comment]
I've started using [Gitako](https://github.com/EnixCoda/Gitako). Works great for me, you can browse all code (not only PRs) and it even has a search feature 😍 

![image](https://user-images.githubusercontent.com/1316517/120507338-02558780-c3c7-11eb-9ff2-1fc7a45207a8.png)

**Comment by jaraqueffdc** [comment]
I am surprised this is not part of PRs yet. I find that PRs in github would benefit from a big overhaul. If you are used to Merge Requests in Gitlab, for example, coming to github for a PR turns into a huge pain. Not because you are used to a different workflow or layout but because there are a lot of things you would expect are standard. The fact that there are many tabs for different things, conversations are hard to follow, no tree view to get a birds view of what changed... I don't know, it seems clunky.

It would be great if we could get at least the tree view to get a quick overview of everything that has been changed.

**Comment by bang-eoh** [comment]
+1 from gitlab

**Comment by timbomckay** [comment]
Azure DevOps also has this feature. Definitely a missing feature in GitHub.

**Comment by pcurtain** [comment]
I'd really love to hear github's response on this.   :-)
Do they have some countering feedback that finds it clearer without the tree structure.  
I can't see why it isn't already in the product.

**Comment by wengyhong** [comment]
Giving this a good bump!

**Comment by Popov-VV** [comment]
Up +100500

**Comment by bogomya** [comment]
Up

**Comment by willvictor** [comment]
FYI- github now has built in VS Code Editor. 

_**If viewing a PR in github, just press "." on your keyboard and it will load github in "vs code mode".**_ 

VS code has a very nice PR review workflow (including a tree view of the changed folder structure). In the vs code view, you can comment on the PR, approve it, etc. 

I'd highly recommend it! As far as i can tell, this feature satisfies the use case described here (but i don't want to speak for everyone).

**Comment by zixinhuangrtsl** [comment]
Upppp

**Comment by mbohgard** [comment]
I think an "UP!" is in order...

**Comment by danieltranca** [comment]
+1

**Comment by hueyyeng** [comment]
+1

After using GitLab at my current workplace, this a sorely missing in GitHub when I need to review multiple files.

**Comment by damien-thiesson** [comment]
+1

**Comment by gonan-gl** [comment]
+1
And too bad that we have to resort to an "external" tool such as integrated vscode to get what has been requested for years.
In the meantime, this integrated vscode really makes up for this lack.

**Comment by zerkowsm** [comment]
+1

**Comment by ndrsllwngr** [comment]
This feature request has now been picked up by GitHub and can be found on the public roadmap[^1]. :tada:
- https://github.com/github/roadmap/issues/348

[^1]: Currently it is in the _Q1 2022 – Jan-Mar_ column of https://github.com/orgs/github/projects/4247/views/1.

**Comment by xserrat** [comment]
> This feature request has now been picked up by GitHub and can be found on the public roadmap[1](#user-content-fn-1-21663a4f0a25486bbfec64c5f265d052). 🎉
> 
> * [Tree view for changed files in a pull request github/roadmap#348](https://github.com/github/roadmap/issues/348)
> 
> ## Footnotes
> 1. Currently it is in the _Q1 2022 – Jan-Mar_ column of https://github.com/orgs/github/projects/4247/views/1. [↩](#user-content-fnref-1-21663a4f0a25486bbfec64c5f265d052)

What great news @ndrsllwngr !! Thanks for your info and thank you to everyone for the support of this feature request 👏🏽

**Comment by ndrsllwngr** [comment]
The **Pull Request File Tree** is now available as [Feature preview](https://docs.github.com/en/get-started/using-github/exploring-early-access-releases-with-feature-preview)
- Feedback https://github.com/github/feedback/discussions/12341

![Screenshot from 2022-03-03 16-22-20](https://user-images.githubusercontent.com/10077258/156595071-1097c9cb-2513-447a-b9eb-8b31f6f2f30a.png)

**Comment by xserrat** [comment]
We have achieved!!🥳🥳🥳 I'm going to close the issue☺️

**Comment by AlexSwensen** [comment]
This is fantastic! Great job!
